libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -0,0 +1,337 @@
+{
+	VirtualLookup const& requiredLookup = *_modifierInvocation.name().annotation().requiredLookup;
+
+	if (auto const* modifier = dynamic_cast<ModifierDefinition const*>(_modifierInvocation.name().annotation().referencedDeclaration))
+	{
+		if (requiredLookup == VirtualLookup::Virtual)
+			modifier = &modifier->resolveVirtual(m_graph.contract);
+		else
+			solAssert(requiredLookup == VirtualLookup::Static, "");
+
+		functionReferenced(*modifier);
+	}
+}
+
+void FunctionCallGraphBuilder::enqueueCallable(CallableDeclaration const& _callable)
+{
+	if (!m_graph.edges.count(&_callable))
+	{
+		m_visitQueue.push_back(&_callable);
+
+		// Insert the callable to the graph (with no edges coming out of it) to mark it as visited.
+		m_graph.edges.insert({Node(&_callable), {}});
+	}
+}
+
+void FunctionCallGraphBuilder::processQueue()
+{
+	solAssert(m_currentNode == Node(SpecialNode::Entry), "Visit queue is already being processed.");
+
+	while (!m_visitQueue.empty())
+	{
+		m_currentNode = m_visitQueue.front();
+		m_visitQueue.pop_front();
+		get<CallableDeclaration const*>(m_currentNode)->accept(*this);
+	}
+
+	m_currentNode = SpecialNode::Entry;
+}
+
+bool FunctionCallGraphBuilder::add(Node _caller, Node _callee)
+{
+	return m_graph.edges[_caller].insert(_callee).second;
+}
+
+void FunctionCallGraphBuilder::functionReferenced(CallableDeclaration const& _callable, bool _calledDirectly)
+{
+	if (_calledDirectly)
+	{
+		solAssert(
+			holds_alternative<SpecialNode>(m_currentNode) || m_graph.edges.count(m_currentNode) > 0,
+			"Adding an edge from a node that has not been visited yet."
+		);
+
+		add(m_currentNode, &_callable);
+	}
+	else
+		add(SpecialNode::InternalDispatch, &_callable);
+
+	enqueueCallable(_callable);
+}
+
+ostream& solidity::frontend::operator<<(ostream& _out, FunctionCallGraphBuilder::Node const& _node)
+{
+	using SpecialNode = FunctionCallGraphBuilder::SpecialNode;
+
+	if (holds_alternative<SpecialNode>(_node))
+	{
+		auto specialNode = get<SpecialNode>(_node);
+		switch (specialNode)
+		{
+			case SpecialNode::InternalDispatch:
+				_out<< "InternalDispatch";
+				break;
+			case SpecialNode::Entry:
+				_out<< "Entry";
+				break;
+			default: solAssert(false, "Invalid SpecialNode type");
+		}
+	}
+	else
+	{
+		auto const* callableDeclaration = get<CallableDeclaration const*>(_node);
+		solAssert(callableDeclaration, "");
+
+		auto const* function = dynamic_cast<FunctionDefinition const *>(callableDeclaration);
+		auto const* event = dynamic_cast<EventDefinition const *>(callableDeclaration);
+		auto const* modifier = dynamic_cast<ModifierDefinition const *>(callableDeclaration);
+
+		auto typeToString = [](auto const& _var) -> string { return _var->type()->toString(true); };
+		vector<string> parameters = callableDeclaration->parameters() | views::transform(typeToString) | to<vector<string>>();
+		string joinedParameters = parameters | views::join(',') | to<string>();
```

Comment:
You could use `joinHumanReadable`

Reply:
I can't chain `joinHumanReadable()` with `|` though :)

But I switch to `joinHumanReadable()` because I had problems with chaining it anyway (see https://github.com/ethereum/solidity/pull/10973#discussion_r578044154).

Reply:
Done.